content/blog/running-self-hosted-agents-in-azure-devops-using-an-azure-ad-access-token.md

```diff
@@ -0,0 +1,126 @@
+
+Next up, go to API Permissions and add the Azure Devops `user_impersonation` Delegated permission. This will allow the Service Principal to call the Azure Devops REST API, which is what the Agent software is doing when registering itself as a new agent inside a given agent pool.
+
+{{<img src="/img/blog/add-api-permissions-to-app-registration.png" class="img-fluid" title="Assign API Permissions" >}}
+
+Afterwards, we need to assign a role to this app registration so that it is bound to a specific subscription. Go to your given subscription and go to Identitiy and Access Management (IAM) in the blade on your left. We're going to add a new Role Assignment, so follow the given dialogue and make sure that you're setting our new SP to the "Reader" role.
+
+{{<img src="/img/blog/add-role-assignment.png" class="img-fluid" title="Add Role Assignment" >}}
+
+If you forgot to do this, during login you'll get the error `No subscriptions found for {ClientID}`, hopefully pointing you towards fixing this by following the above.
+
+{{<img src="/img/blog/no-role-assignment-error.png" class="img-fluid" title="No Role Assignment" >}}
+
+Of course, just giving the Service Principal permission to call the ADO API is not enough to do what we want to do here, we also have to give the SP the right permissions inside ADO.
+
+First of all, navigate to the Users tab for your organization and add the service principal as a user with Basic permissions. We're going to make it a Project Contributor on the Azure Devops Project I've created earlier, conveniently named "SelfHostedAgent".
+
+{{<img src="/img/blog/add-sp-as-ado-user.png" class="img-fluid" title="Add Service Principal as ADO user" >}}
+
+Next up, we're going to navigate to the Agent Pool we've already created and make sure that the Service Principal (which is now treated the same as an ADO user) can actually Read & Manage the agent pool. For this, we need to add the User Permissions at the Organization Level and add our Service Principal as an Administrator.
+
+Why at the Organization level? If you set the Service Principal as an Administrator at the Project level, the agent will start up but it will give you an error message that somewhat resembles `{Service Principal} needs Manage permissions for pool {Pool Name} to perform the action` and then just error out. This probably has something to do with the fact that Agent Pools are actually an organization-level construct in ADO that are just shared between different projects based on permissions.
+
+{{<img src="/img/blog/set-agent-pool-user-permissions.png" class="img-fluid" title="Set Agent Pool user permissions" >}}
+
+After this you should be set! To be sure, let's just run a little script to make sure that we've got the basics down so far.
+
+```bash
+#!/usr/bin/env bash
+
+set -euo pipefail
+
+readonly ORGANIZATION="YOUR_ORG_HERE"
+readonly POOL_ID="YOUR_POOL_ID"
+readonly API_VERSION="7.1-preview.1"
+
+TOKEN=$(az account get-access-token \
```

Comment:
technically you'd want `TOKEN` to be readonly as well 😉

Reply:
I guess that makes sense, yes